### ISSUE TITLE
Add a well-known format description for ISO 8601 basic format

### ISSUE THREAD

**Issue by korrat** [time-rs/time#378]
In a recent project, I needed to parse date times in the ISO 8601 format (`YYYYMMDDTHHMMSSZ`). I managed to implement this using a format description, but this was quite hacky and will only work for UTC dates using the `Z` time zone specifier.

Do you think this would be a worthwhile addition? I can provide a PR if you're willing to consider it.

**Comment by jhpratt** [comment]
Unfortunately ISO 8601 costs hundreds of dollars. If there's some way to verify the exact format without having the full specification, that would be great.

While I know that the format you've provided is _generally_ correct, that isn't sufficient. There are certainly nuances that need to be handled when parsing, which is impossible to do without the specification itself. That said, if there's some company that wants to purchase a copy of the specification (ISO 8601-1:2019 and ISO 8601-2:2019), I'd be happy to move further towards spec compliance.

**Comment by BartMassey** [comment]
[Ask ISO](https://www.iso.org/contact-iso.html) to donate you a copy of the ISO 8601 documents. Seriously. I'm pretty sure they'll say yes: it's in their interests to have the Rust time crate implement this stuff.

I hate to be part of their pay-to-use standards process, but it's probably the right way forward here.

**Comment by jhpratt** [comment]
I'd never considered that, to be honest. I'll reach out :+1:

**Comment by jhpratt** [comment]
Given the lack of a response, I presume they're not interested.

**Comment by BartMassey** [comment]
Frustrating. Now ask the Rust Foundation to buy these and share them with you, I think.

**Comment by jhpratt** [comment]
Marking as blocked until I am able to get a copy of the standard.

**Comment by nmathewson** [comment]
Would it be adequate to implement the profile of ISO8601 described in RFC 3339?

**Comment by jhpratt** [comment]
@nmathewson RFC3339 is already implemented. This issue is a request for the full spec of ISO 8601.

**Comment by huxi** [comment]
The [Wikipedia article](https://en.wikipedia.org/wiki/ISO_8601#Combined_date_and_time_representations) contains all info required for the ticket.

This is the relevant part of the 2004 version of the spec:

> The following are examples of complete representations of date and time of day representations:

```
Type     | Format                    | Example
+++++++++|+++++++++++++++++++++++++++|++++++++++++++++++++++++++
Basic    | YYYYMMDDThhmmss           | 19850412T101530
Basic    | YYYYMMDDThhmmssZ          | 19850412T101530Z
Basic    | YYYYMMDDThhmmss±hhmm      | 19850412T101530+0400
Basic    | YYYYMMDDThhmmss±hh        | 19850412T101530+04
Extended | YYYY-MM-DDThh:mm:ss       | 1985-04-12T10:15:30
Extended | YYYY-MM-DDThh:mm:ssZ      | 1985-04-12T10:15:30Z
Extended | YYYY-MM-DDThh:mm:ss±hh:mm | 1985-04-12T10:15:30+04:00
Extended | YYYY-MM-DDThh:mm:ss±hh    | 1985-04-12T10:15:30+04
```

The `Z` at the end of the datetimes is the UTC designator.

Also see [Times](https://en.wikipedia.org/wiki/ISO_8601#Times) section on WP regarding fraction of seconds since both `.` and `,` are allowed as separators.

[RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339) and ISO 8601 are mostly identical The main difference is that ISO 8601 would like to have `Z` in case of UTC while `+00:00` would also be ok.

What this issue requested was support for the `Basic` formats in the table above which are only defined in ISO 8601 but not in RFC 3339 because that demands the extended format.

Hope this helps.

**Comment by jhpratt** [comment]
What I'm looking for is the 2019 version, not the 2004 version. I have no idea if there are differences as information is hard to come by. I'm not going to rely on Wikipedia regardless; there are frequently nuances that need to be handled that are undoubtedly not shown in a high-level article. That's why I need the actual specification.

**Comment by huxi** [comment]
There's [this interesting discussion](https://twitter.com/isostandards/status/1367138676162105344) about the topic on Twitter, including [this nugget](https://twitter.com/punycode/status/1369070121327726592).

The thing is that even if we would fund you a copy of ISO 8601-1:2019, you couldn't really use it for anything in an open source project since ISO would like to charge **per reader** and has a very restrictive license. The situation would just change from me telling you "Yeah, it looks like this. Trust me." to you telling other maintainers "Yeah, it looks like this. Trust me.".

You couldn't reference the standard in a meaningful way since no one would be able to check those references. The situation is indeed kafkaesque.

It all boils down to ISO 8601 implementations always being hearsay and best-effort - as enforced by ISO legalese itself.

For "ISO 8601 datetime (basic)" this hearsay is "RFC 3339 format, but without `-` and `:`" and "`Z` is equivalent to `+00:00`".

I do have a legal copy of the previous version because a former employer bought a copy for me. Since those copies are **per reader**, I assume that it's my own personal copy and not the copy of my former employer. I assure you that the information I provided above is all the information provided in the text. The specification contains more formats but this is the part concerned with datetime.

The spec also contains this copyright notice:

> All rights reserved. Unless otherwise specified, no part of this publication may be reproduced or utilized in any form or by any means, electronic or mechanical, including photocopying and microfilm, without permission in writing from either ISO at the address below or ISO's member body in the country of the requester.

IANAL, but this seems to make it illegal to even use the examples provided in the spec in publicly available test code. This also applies to the examples I provided above.

Yes. This is all very, very absurd.

I'm not trying to push you into any direction here. I only wanted to give you the full picture.

**Comment by jhpratt** [comment]
My knowledge of copyright is that while the text of the spec can be copyrighted, the format itself cannot be, and thus I would be able to have things like tests public. I'd just have to create my own examples...which is not a big deal. I do understand everything is ultimately hearsay as you've mentioned.

I think what I'm going to do is reach out to both the Free Software Foundation and the Rust Foundation to see if either or both is able to provide legal guidance on this front.

**Comment by huxi** [comment]
Yeah, creating own examples for tests would be trivial but the opposite of what one would normally do while implementing a spec. One would usually use exactly the examples from the spec because they may contain all kinds of edge-cases.
Most of the spec is just that: patterns with examples. And those patterns are copyrighted, too. 😂

The situation is just extraordinarily absurd. And it has been exactly like that for 20 years.

Every company I ever worked for has suffered in some way because of this.
The problem isn't the price of the spec. The problem is that you can't just send someone a link and say "This is what it should look like. And the authority for this demand is ISO 8601.". It causes unnecessary friction which equals cost. Sometimes lots of it.

**Comment by jhpratt** [comment]
That's where I'm hoping some foundation can have an attorney clarify for me what is and is not permissible. If it does turn out that it's largely not possible to do this, then this issue will unfortunately have to be closed for legal reasons. I'll be reaching out either later today or tomorrow most likely.

**Comment by jhpratt** [comment]
For transparency I reached out to the Rust Foundation, the Free Software Foundation, and the Electronic Frontier Foundation mid-afternoon (US ET) on January 3. I have not yet received a response. If there is no response by the end of next week (that is, by the end of the day January 14), I will close this issue. with no guidance whatsoever, even from the very foundation that should theoretically help library maintainers in Rust, there won't be anything I can reasonably do.

**Comment by BartMassey** [comment]
Seems reasonable. Thanks for looking into it so thoroughly.

One other possible way to proceed would be to implement @huxi's "For 'ISO 8601 datetime (basic)' this hearsay is 'RFC 3339 format, but without - and :' and 'Z is equivalent to +00:00'" as "ISO8601ish format (not known to be True ISO8601™)." I wouldn't be surprised if everyone would be perfectly happy with that…

**Comment by jhpratt** [comment]
Closing as no response has been received from any of the three foundations I contacted.

**Comment by jhpratt** [comment]
Over the holiday weekend I received the following response (in part) from the Rust Foundation:

> Our understanding is that the license text that you have quoted applies to the specification document only, not the standard itself, and that it does not restrict implementation of the standard.
>
> Therefore, you should be able to conduct the activities you suggest. 

The Free Software Foundation also responded, but deemed my question to be too close to legal advice, and as such did not provide guidance.

With the response from the Rust Foundation, I'm satisfied that proceeding would not violate the license so long as everything written is my own (including creating my own tests). So at this point it's waiting on a company to provide me with a copy of the specification itself.

**Comment by ColmanHumphrey** [comment]
@jhpratt is [this the right link](https://www.iso.org/iso-8601-date-and-time-format.html)?  and do you need part 1 and part 2?

We at [Slight](https://www.slight.co/) can definitely support half this (or e.g. one of the parts). We'd love to go half with another company. Let me see if I can find another (@jeffwilcox maybe). To be clear, we can do part 1 unconditionally (or part 2 if you prefer) even if we can't find a partner.

**Comment by jhpratt** [comment]
Hey @ColmanHumphrey! Thanks for the offer. Yes, that's the correct link. After quickly reading, part 1 appears to be what I need to satisfy this request. Part 2 _could_ be useful for some other things, but I'm not certain where that would fit in with my vision for this crate.

I know you've got my email as you sent one just a few days ago. No rush or anything — I have some other things I'm working on in the compiler at the moment. Thanks again for offering to cover this.

**Comment by ColmanHumphrey** [comment]
Sounds good @jhpratt! I've just emailed the ISO org to find out how to buy it in your name, so it'll probably take a bit to sort that out anyway.

**Comment by jhpratt** [comment]
Thanks to @ColmanHumphrey & @SlightCode, I am now in possession of ISO 8601-1:2019. It's likely that I will look over the document and begin implementation in the coming weeks.

**Comment by jhpratt** [comment]
Finally have a bit of time to dig through the formats in the specification. There is quite a bit of flexibility.

 There is the possibility of configuration for some things. Some of these:
- whether a format is "basic" or "extended"
- the exact components present (date, time, and offset) — can be done in any combination except date-offset
- whether the date is the calendar date, week date, or ordinal date

Since it appears as though a number of people are following this issue, I'll ask here. **Do users need the ability to configure this at runtime?** It should only affect formatting, as the parser should be able to cope with most if not all of the flexibility permitted (I'll know for sure when I start implementation).

If the ability to configure this at runtime isn't needed, I can use const generics to reduce the size of the struct. This'll likely need a bit of creativity to work around current const generics restrictions, but it's doable without too much quirkiness.

**Comment by mati865** [comment]
> Do users need the ability to configure this at runtime? It should only affect formatting, as the parser should be able to cope with most if not all of the flexibility permitted (I'll know for sure when I start implementation).

Hello, I'm one of developers in quite big corporation. In our use case being able to parse ISO 8601 is high on nice to have list.
Formatting on the other hand is more like whatever as long as its compilant with the standard.

**Comment by jhpratt** [comment]
Like I said, parsing should be able to handle any valid ISO 8601 regardless of configuration, as the format appears to be designed specifically to avoid any ambiguities. Formatting I intend on providing options to the end user. I'm just not certain whether there's a use case for dynamically setting the configuration values (and if it's significant enough to outweigh the benefits of ZST).

**Comment by BartMassey** [comment]
As a non-client of this functionality, most likely, my two cents is that it is better to start with the more straightforward runtime-configured struct. I have a hard time imagining the use case for a smaller format struct; I can imagine cases where for example different services accessed by the same client would want different formats; const generics are still in-progress and are harder for newer Rust programmers to understand and work with.

**Comment by korrat** [comment]
In my applications, I do not see the need to provide runtime configurable formatting at the moment. If it became relevant in the future, I think the existing formatting facilities are sufficiently powerful to handle that as well.

So, no, I don't think that runtime configuration is necessary.

**Comment by malobre** [comment]
I believe a runtime configuration would be more flexible. For example in a web API, you might retrieve a date-time from a DB and use it with various other APIs that might have different format requirements.

**Comment by jhpratt** [comment]
Please re-read my original comment. I am seeking specific real-world use cases where runtime configuration is necessary for _formatting_. Parsing appears to be workable regardless of configuration, as the format is unambiguous.

Further comments similar to ones already made will be outright deleted, as I've been quite clear in the type of use case I'm looking for.

**Comment by jhpratt** [comment]
At this point, I have both parsing and formatting fully implemented. I am writing numerous tests and ensuring 100% code coverage before pushing everything up. But it's happening :tada:

**Comment by jhpratt** [comment]
I know there's a number of people following this, so I'll try to provide an update. Testing is going well, and led me to catch 3 errors so far. It gets quite tedious at times, so I'm not putting as much time into it as I theoretically could. With that said, over the coming days, I have family things and will then be taking a lengthy flight. At that point I'll be spending some time in a new country for work. I'm sure I'll do some exploring in my free time, as I am only there for a limited time. That will take priority over this (sorry not sorry). Once I return home I've got a little more going on, but will be back to "normal" by June 6. So I wouldn't _plan_ on having this completed before then. At the absolute earliest, this will be released on June 2. That is when the MSRV can become Rust 1.57 per the crate's six-month policy, and the implementation of ISO 8601 requires const panics to work.

TL;DR get your hopes up, but let's be realistic with a timeframe :slightly_smiling_face: Next release on June 2 at the earliest, realistically June 6 or soon thereafter.

**Comment by jhpratt** [comment]
I have just pushed ISO 8601 support to main. It's not as good as it will be in the future, but it fully complies to the standard and the public API is final.

The two major things I'd like to improve on are performance and errors. The errors returned from failed parsing are terrible — technically correct but wildly misleading. Incidentally, the manner in which I will improve performance will also improve the error messages (it's effectively specialization based on the target type).

Please take a look at the implementation and give it a shot! There's a couple unrelated things I want to do before cutting a release, so there'll be a few days at the least before it goes out. I'm 100% abusing const generics, but I think it's ergonomic enough that it shouldn't be an issue.

Thanks again to @ColmanHumphrey and @slightcode for purchasing the standard in my name. This wouldn't have happened otherwise.

**Comment by ColmanHumphrey** [comment]
Nice one @jhpratt, great to see this!
